src/Npgsql/Internal/Converters/Primitive/PgNumeric.cs

```diff
@@ -356,7 +356,7 @@ internal static decimal ToDecimal(short scale, short weight, ushort sign, Span<s
                     while (scaleDifference > 0)
                     {
                         var scaleChunk = Math.Min(MaxUIntScale, scaleDifference);
-                        result *= UIntPowers10[scaleChunk];
+                        scaleFactor *= UIntPowers10[scaleChunk];
```

Comment:
Nice fix. Do you think we could just change scaleFactor to be adjusted correctly at the start? We may be able to remove this entire block.

Reply:
Looking at line 354 btw, that seems like it can very easily throw out of bounds if the scaleDifference is too big. Not sure why it's not done in a similar chunking loop as the positive side.

Reply:
> Do you think we could just change scaleFactor to be adjusted correctly at the start? We may be able to remove this entire block.

We probably can, though it seems to be extremely non-straighforward. For example just by changing `result /= UIntPowers10[-scaleDifference]` to `scaleFactor /= UIntPowers10[-scaleDifference]` (which theoretically should be the same) results in us potentially adding a few zeroes (as in getting 1.2000 instead of 1.2).

It might be better to instead try to reimplement `PgNumeric.ToDecimal` to work via decimal's constructor (`lo`, `mid` and `hi` bytes). But that's definitely not something we should backport to 8.0.

> Looking at line 354 btw, that seems like it can very easily throw out of bounds if the scaleDifference is too big. Not sure why it's not done in a similar chunking loop as the positive side.

From my experiments it didn't seem possible due to `weight` or `scale` always compensating for `digitScale`. But I did add `while` loop just in case.

Reply:
>From my experiments it didn't seem possible due to weight or scale always compensating for digitScale. But I did add while loop just in case.

Thanks. I know since pg 15 scale can be negative too, so we might be able to create a test case for it. I'm not sure if this is something that is sent or handled on the server side as it's a rounding instruction.

>Beginning in PostgreSQL 15, it is allowed to declare a numeric column with a negative scale. Then values will be rounded to the left of the decimal point. The precision still represents the maximum number of non-rounded digits.

https://www.postgresql.org/docs/current/datatype-numeric.html

Given this description, if it's at all sent client side we don't support this correctly, even if we don't crash.

Reply:
>We probably can, though it seems to be extremely non-straighforward. For example just by changing result /= UIntPowers10[-scaleDifference] to scaleFactor /= UIntPowers10[-scaleDifference] (which theoretically should be the same) results in us potentially adding a few zeroes (as in getting 1.2000 instead of 1.2).

Yes there are some asymmetries around negative scale, so we don't actually have a scaleFactor of the supplied scale when scale is < 0.

>It might be better to instead try to reimplement PgNumeric.ToDecimal to work via decimal's constructor (lo, mid and hi bytes). But that's definitely not something we should backport to 8.0.

The conversion work here really is mapping the numeric digits, sent as shorts limited to a power of 10 amount (i.e. 10k though technically configurable), to the 3 decimal uints. Not sure it gets much easier when we do the math of spilling results into the next uint ourselves, and it'll be just as iterative.

Reply:
> Thanks. I know since pg 15 scale can be negative too, so we might be able to create a test case for it. I'm not sure if this is something that is sent or handled on the server side as it's a rounding instruction

Just checked, in case of negative scale pg truncates the value on its side, so we're in the clear.